### PR TITLE
CSHARP-2786: Verify that PredicateTranslatorValidationTests pass with LINQ3.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs [hunk 1]
```diff
@@ -67,14 +67,15 @@ public void Should_not_throw_the_exception_when_a_child_expression_is_called_fro
...
 
-            Expression<Func<TestObject, bool>> expr = (a) => a.Collection1.Any(b => local.Any(c => b.Collection2.Contains(c)));
-            AssertWhere(expr, "{ $match : { 'Collection2' : { '$in' : [1, 2] } } }");
+            Expression<Func<TestObject, bool>> expr;
+            expr = (a) => a.Collection1.Any(b => local.Any(c => b.Collection2.Contains(c)));
+            AssertWhere(expr, "{ $match : { Collection1 : { $elemMatch : { Collection2 : { $in : [1, 2] } } } } }");
 
             expr = (a) => a.Collection1.Any(b => b.Collection1.Any(c => local.Any(d => c.Collection2.Contains(d))));
-            AssertWhere(expr, "{ $match : { 'Collection2' : { '$in' : [1, 2] } } }");
+            AssertWhere(expr, "{ $match : { Collection1 : { $elemMatch : { Collection1 : { $elemMatch : { Collection2 : { $in : [1, 2] } } } } } } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs:75]: As Dmitry questioned in CSHARP-2786 LINQ2 was not translating this query correctly.

File: tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs [hunk 2]
```diff
@@ -192,56 +193,49 @@ public void Should_not_throw_the_exception_when_top_method_is_not_where_predicat
...
-        public void Should_throw_the_exception_when_a_child_expression_uses_parameters_from_grandparents()
+        public void Should_not_throw_when_a_child_expression_uses_parameters_from_grandparents()
         {
             Setup();
 
             Expression<Func<TestObject, bool>> expr = (a) => a.Collection1.Any(b => b.Collection1.Any(c => a.Value1 == 2));
 
-            var exception = Assert.Throws<NotSupportedException>(() => Execute(CreateWhereQuery(expr)));
-            exception.Message.Should().Be(string.Format(NotSupportErrorMessageTemplate, "{document}{Collection1}.Where(Any({document}{Collection1}.Where(({document}{Value1} == 2))))", "a"));
+            AssertWhere(expr, "{ $match : { $expr : { $anyElementTrue : { $map : { input : '$Collection1', as : 'b', in : { $anyElementTrue : { $map : { input : '$$b.Collection1', as : 'c', in : { $eq : ['$Value1', 2] } } } } } } } } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs:202]: A lot of queries that LINQ2 threw a not supported exception on can be handled in LINQ3 using `$expr`.

**Review by rstam — COMMENTED** [review]
The main job when reviewing this PR is to verify that the generated MQL is correct.

The generated MQL is quite deeply nested in some cases so this might not be easy.

I checked it all and believe it to be correct, but it needs a second set of eyes.

**Comment by rstam** [comment]
I went off on the wrong track because I was editing the `PredicateTranslatorValidationTests.cs` file in the `Linq2ImplementationTests` folder, which I should not have done because this file is testing LINQ2.

If you do a diff against the `PredicateTranslatorValidationTests.cs` file in the `Linq2ImplementationTestsOnLinq3` folder you will find that almost exactly these same changes were made once before.

Based on that I think this PR should not be pushed to master and CSHARP-2786 can just be closed as "works in LINQ3".

File: tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs (hunk 2 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Tests/Linq/Linq2ImplementationTests/Translators/PredicateTranslatorValidationTests.cs:202]: Nice!

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Comment by rstam** [comment]
Closed without merging to master.

**Pull request closed without merging**